commit a027f56aa89e9cdfed57ee8d96f3420cab204d5d
Author: Albert Astals Cid <aacid@kde.org>
Date:   Tue Apr 16 19:58:48 2019 +0200

    Fix stack-buffer-overflow read on malformed files
    
    Summary:
    Make sure we read at most 32 bytes for the user/group
    
    oss-fuzz #14264
    
    Reviewers: apol
    
    Reviewed By: apol
    
    Subscribers: kde-frameworks-devel
    
    Tags: #frameworks
    
    Differential Revision: https://phabricator.kde.org/D20604

diff --git a/src/ktar.cpp b/src/ktar.cpp
index c67c079..398e5ac 100644
--- a/src/ktar.cpp
+++ b/src/ktar.cpp
@@ -421,8 +421,13 @@ bool KTar::openArchive(QIODevice::OpenMode mode)
             int access = strtol(p, &dummy, 8);
 
             // read user and group
-            QString user = QString::fromLocal8Bit(buffer + 0x109);
-            QString group = QString::fromLocal8Bit(buffer + 0x129);
+            const int maxUserGroupLength = 32;
+            const char *userStart = buffer + 0x109;
+            const int userLen = qstrnlen(userStart, maxUserGroupLength);
+            const QString user = QString::fromLocal8Bit(userStart, userLen);
+            const char *groupStart = buffer + 0x129;
+            const int groupLen = qstrnlen(groupStart, maxUserGroupLength);
+            const QString group = QString::fromLocal8Bit(groupStart, groupLen);
 
             // read time
             buffer[0x93] = 0;
